core/src/types.ts

```diff
@@ -76,22 +81,42 @@ export declare interface AuctionTransaction extends Auction, TransactionFees {
     combinedSwapFeesETH: BigNumber;
 }
 
-export declare interface RegularCalleeConfig {
+declare interface Routed {
+    route: string[];
+}
+
+declare interface Tokenized {
+    token0: string;
+    token1: string;
+}
+
+export declare interface RegularCalleeConfig extends Routed {
     callee:
         | 'UniswapV2CalleeDai'
         | 'WstETHCurveUniv3Callee'
         | 'CurveLpTokenUniv3Callee'
         | 'UniswapV3Callee'
         | 'rETHCurveUniv3Callee';
-    route: string[];
 }
 
-export declare interface UniswapV2LpTokenCalleeConfig {
+export declare interface UniswapV2LpTokenCalleeConfig extends Tokenized {
     callee: 'UniswapV2LpTokenCalleeDai';
-    token0: string;
-    token1: string;
 }
 
+declare interface MarketDataBase {
+    marketUnitPrice: BigNumber;
+    marketUnitPriceToUnitPriceRatio?: BigNumber;
+    transactionGrossProfit?: BigNumber;
+    transactionGrossProfitDate?: Date;
+    transactionNetProfit?: BigNumber;
+}
+
+declare interface MarketDataRouted extends MarketDataBase, Routed {}
+
+declare interface MarketDataTokenized extends MarketDataBase, Tokenized {}
+
+export type MarketData = MarketDataRouted | MarketDataTokenized;
```

Comment:
Why did you have to create new `Routed` type and extend it instead of `RegularCalleeConfig`? That is the problem you're experiencing here: https://github.com/sidestream-tech/unified-auctions-ui/pull/511#discussion_r1009555381

When you check callee name of the `RegularCalleeConfig`-type object, typescript can already say if `route` or `token*` keys should be present. But when you do the same for `MarketData` there is no (typed) connection between the callee name and the keys.

```suggestion
declare interface MarketDataRegular extends MarketDataBase, RegularCalleeConfig {}

declare interface MarketDataUniswapV2Lp extends MarketDataBase, UniswapV2LpTokenCalleeConfig {}

export type MarketData = MarketDataRegular | MarketDataUniswapV2Lp;
```

Reply:
Why exactly it did work before when it was in the `exchange` object directly `exchange.callee`, but doesn't work when it's one level deeper, in the `exchanges` record `exchanges[exchange].callee`?

Reply:
> since we know for certain that the `callee` name narrows down the type

No, it's not optimal. The linked issue doesn't talk about nesting at all. As I said, `exchange.callee` check did work before (so it's not a typescript problem), but currently types are implemented differently and with the current implementation I don't even expect it to work. Have you tried [the suggestion](https://github.com/sidestream-tech/unified-auctions-ui/pull/511#discussion_r1009568739)? Where can I see a commit with it?